### DISCUSSION TITLE
How to configure rustfmt

### DISCUSSION THREAD

**Discussion by imding** [zed-industries/zed#6595] [Feature Requests]
Hi, I am trying out Zed and was wondering if I could configure rustfmt when using `cmd + shift + i`.

**Comment by JosephTLyons** [comment]
I don't *think* we currently have a way to associate commands with a given key binding, but I've heard others mentiong wanting this, so we might want to convert this into a feature request.

**Reply by johnm** [reply]
Hm... Is there some reason Zed's format on save is not respecting the rustfmt.toml file in a Rust project directory?

**Reply by marcospb19-cw** [reply]
Bumping, any workaround for the formatter to respect `rustfmt.toml`?
Or any open issues for it?

**Reply by UnknownSuperficialNight** [reply]
Bumping, any updates on this?

**Reply by marcospb19-cw** [reply]
For me, the problem was that `zed` wasn't running `rustfmt` in `nightly`.

I solved it with the suggestion at https://github.com/zed-industries/zed/issues/4491#issuecomment-1945657731:

```json
"lsp": {
  "rust-analyzer": {
    "initialization_options": {
      "rustfmt": {
        "extraArgs": ["+nightly"]
      }
    }
  }
}
```

I thought I had to configure in `Zed`, but it's reading fine from `rustfmt.toml`.

So, I guess the answer is: just edit the `rustfmt.toml` like usual.

**Reply by imding** [reply]
are you saying Zed **is** respecting your `rustfmt.toml` by default except when you're running nightly version of Zed?

for me, I'm not running nightly version of Zed and it still doesn't respect `rustfmt.toml`

**Reply by marcospb19-cw** [reply]
No, I'm not running nightly version of Zed, instead, my `rustfmt.toml` file had some settings that require nightly.

> for me, I'm not running nightly version of Zed and it still doesn't respect rustfmt.toml

Can you provide a small project example where `cargo fmt` respects `rustfmt.toml` but when formatting in `Zed`, it (Rust-Analyzer) doesn't?

With that we'd be able to create an issue.

**Reply by imding** [reply]
oh I understand now, it turns out the options in my `.rustfmt.toml` are unstable.

**Reply by ekakshjanweja** [reply]
hey, i am new to rust,
i am using the standard version of rustfmt, how do I get format on save to work, I am not using the nightly version of the fmt

**Comment by imding** [comment]
my solution was to follow [the rustfmt repo instruction](https://github.com/rust-lang/rustfmt?tab=readme-ov-file#on-the-nightly-toolchain).

you may get
```bash
error: toolchain 'nightly-...' is not installed
```

in which case, run `rustup toolchain install nightly-...` then run `cargo +nightly fmt` from project root to test that it works.

then you can add the 
```json
"lsp": {
  "rust-analyzer": {
    "initialization_options": {
      "rustfmt": {
        "extraArgs": ["+nightly"]
      }
    }
  }
}
```
settings to Zed.

**Comment by kantuni** [comment]
IIUC, `rust-analyzer` is using `rustfmt` underneath.
```json
"Rust": {
  "formatter": "language_server",
  "format_on_save": "on"
}
```

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
